### PR TITLE
Fixed compatibility with UE 4.22.3

### PR DESCRIPTION
**Pull request by Scienziatogm**
• This plugin is now compatible with Unreal Engine 4.22.3
• Fixed an issue when GEngine->Exec was not called, and added the chance to run GameMode custom exec functions from the server window

### CODE REVIEW

File: Source/DedicatedServer/Private/DedicatedServer.cpp [hunk 1]
```diff
@@ -29,15 +29,15 @@ void FDedicatedServerModule::StartupModule()
 
 			if( bUseConsole && !TIsSame<decltype( GLogConsole ), FServerConsole>::Value )
 			{
-				g_pConsole = new FServerConsole();
+				g_pConsole = TUniquePtr<FServerConsole>(new FServerConsole());
```
Comment (neico) [inline Source/DedicatedServer/Private/DedicatedServer.cpp]:
```suggestion
				g_pConsole = TUniquePtr<FServerConsole>( new FServerConsole() );
```

File: Source/RCon/RCon.Build.cs [hunk 2]
```diff
@@ -5,8 +5,8 @@
 
 public class RCon : ModuleRules
 {
-	public RCon( TargetInfo Target )
-	{
+	public RCon(ReadOnlyTargetRules Target) : base(Target)
```
Comment (neico) [inline Source/RCon/RCon.Build.cs]:
```suggestion
	public RCon( ReadOnlyTargetRules Target ) : base( Target )
```

File: Source/RCon/RCon.Build.cs [hunk 3]
```diff
@@ -5,8 +5,8 @@
...
+    {
```
Comment (neico) [inline Source/RCon/RCon.Build.cs]:
```suggestion
	{
```

File: Source/DedicatedServer/DedicatedServer.Build.cs [hunk 4]
```diff
@@ -5,12 +5,12 @@
 
 public class DedicatedServer : ModuleRules
 {
-	public DedicatedServer( TargetInfo Target )
-	{
+	public DedicatedServer(ReadOnlyTargetRules Target) : base(Target)
```
Comment (neico) [inline Source/DedicatedServer/DedicatedServer.Build.cs]:
```suggestion
	public DedicatedServer( ReadOnlyTargetRules Target ) : base( Target )
```

File: Source/DedicatedServer/DedicatedServer.Build.cs [hunk 5]
```diff
@@ -5,12 +5,12 @@
...
+    {
```
Comment (neico) [inline Source/DedicatedServer/DedicatedServer.Build.cs]:
```suggestion
	{
```

File: Source/DedicatedServer/Public/ServerConsole.h [hunk 6]
```diff
@@ -6,19 +6,20 @@ DECLARE_LOG_CATEGORY_EXTERN( LogServerConsole, Log, All );
...
-		#include "Private/Linux/LinuxPlatformOutputDevicesPrivate.h"
-		#define FOutputDeviceConsolePlatform FOutputDeviceConsoleLinux
+		#include "Public/Linux/LinuxConsoleOutputDevice.h"
+		#define FOutputDeviceConsolePlatform FLinuxConsoleOutputDevice
 	#else
 		#error How did you get here?
 	#endif
 
-	class FServerConsole : public FOutputDeviceConsolePlatform
+	class FServerConsole : public FOutputDeviceConsole
```
Comment (neico) [inline Source/DedicatedServer/Public/ServerConsole.h]: If you remove `FOutputDeviceConsolePlatform` here then you might as well remove it generally, at a glance it doesn't look like I've used anything that isn't present in `FOutputDevice`

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 7]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+							::WriteConsole(m_hOutputHandle, sOutput, FCString::Strlen(sOutput), &iCharsWritten, NULL);
+
+							m_hCommandHistory.Add(tmp);
+						}
+						else
+						{
+							FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *tmp, LINE_TERMINATOR );
+							::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
+						}
+					});
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
					} );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 8]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
-					else
-					{
-						FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *m_sInput, LINE_TERMINATOR );
-						::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
-					}
+					auto& tmp = m_sInput;
+					AsyncTask(ENamedThreads::GameThread, [this, tmp]() {
+						// code to execute on game thread here
+
+						TCHAR sOutput[MAX_SPRINTF] = TEXT("");
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
						TCHAR sOutput[MAX_SPRINTF] = TEXT( "" );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 9]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
-
-						m_hCommandHistory.Add( m_sInput );
-					}
-					else
-					{
-						FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *m_sInput, LINE_TERMINATOR );
-						::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
-					}
+					auto& tmp = m_sInput;
+					AsyncTask(ENamedThreads::GameThread, [this, tmp]() {
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
					AsyncTask( ENamedThreads::GameThread, [this, sInput]()
					{
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 10]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
-						::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
-
-						m_hCommandHistory.Add( m_sInput );
-					}
-					else
-					{
-						FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *m_sInput, LINE_TERMINATOR );
-						::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
-					}
+					auto& tmp = m_sInput;
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
					auto &sInput = m_sInput;
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 11]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+					AsyncTask(ENamedThreads::GameThread, [this, tmp]() {
+						// code to execute on game thread here
+
+						TCHAR sOutput[MAX_SPRINTF] = TEXT("");
+						unsigned long iCharsWritten;
+
+						if( tmp.StartsWith( TEXT( "help" ) ) )
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
						if( sInput.StartsWith( TEXT( "help" ) ) )
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 12]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						{
+							DumpConsoleHelp();
+
+							FCString::Sprintf( sOutput, TEXT( "> %s%s" ), *tmp, LINE_TERMINATOR );
+							::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
+
+							m_hCommandHistory.Add( tmp );
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							m_hCommandHistory.Add( sInput );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 13]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						// code to execute on game thread here
+
+						TCHAR sOutput[MAX_SPRINTF] = TEXT("");
+						unsigned long iCharsWritten;
+
+						if( tmp.StartsWith( TEXT( "help" ) ) )
+						{
+							DumpConsoleHelp();
+
+							FCString::Sprintf( sOutput, TEXT( "> %s%s" ), *tmp, LINE_TERMINATOR );
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							FCString::Sprintf( sOutput, TEXT( "> %s%s" ), *sInput, LINE_TERMINATOR );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 14]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+							::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
+
+							m_hCommandHistory.Add( tmp );
+						}
+						else if( GEngine->Exec( GWorld, *tmp ) )
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
						else if( GEngine->Exec( GWorld, *sInput ) )
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 15]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						{
+							FCString::Sprintf( sOutput, TEXT( "> %s%s" ), *tmp, LINE_TERMINATOR );
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							FCString::Sprintf( sOutput, TEXT( "> %s%s" ), *sInput, LINE_TERMINATOR );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 16]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+							::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
+
+							m_hCommandHistory.Add( tmp );
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							m_hCommandHistory.Add( sInput );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 17]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						}
+						else if ( GWorld->GetAuthGameMode() != nullptr && GWorld->GetAuthGameMode()->ProcessConsoleExec(*tmp, *GLog, nullptr) )
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
						else if ( GWorld->GetAuthGameMode() != nullptr && GWorld->GetAuthGameMode()->ProcessConsoleExec( *sInput, *GLog, nullptr ) )
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 18]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						{
+							FCString::Sprintf(sOutput, TEXT("> %s%s"), *tmp, LINE_TERMINATOR);
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							FCString::Sprintf( sOutput, TEXT("> %s%s"), *sInput, LINE_TERMINATOR );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 19]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+							::WriteConsole(m_hOutputHandle, sOutput, FCString::Strlen(sOutput), &iCharsWritten, NULL);
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 20]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+
+							m_hCommandHistory.Add(tmp);
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							m_hCommandHistory.Add( sInput );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 21]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
+						}
+						else
+						{
+							FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *tmp, LINE_TERMINATOR );
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
							FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *sInput, LINE_TERMINATOR );
```

File: Source/DedicatedServer/Private/ServerConsoleWindows.cpp [hunk 22]
```diff
@@ -45,31 +47,42 @@ void DumpConsoleHelp();
...
-					{
-						FCString::Sprintf( sOutput, TEXT( "Unknown Command: %s%s" ), *m_sInput, LINE_TERMINATOR );
-						::WriteConsole( m_hOutputHandle, sOutput, FCString::Strlen( sOutput ), &iCharsWritten, NULL );
-					}
+					auto& tmp = m_sInput;
+					AsyncTask(ENamedThreads::GameThread, [this, tmp]() {
+						// code to execute on game thread here
+
+						TCHAR sOutput[MAX_SPRINTF] = TEXT("");
+						unsigned long iCharsWritten;
```
Comment (neico) [inline Source/DedicatedServer/Private/ServerConsoleWindows.cpp]:
```suggestion
						unsigned long iCharsWritten = 0;
```
It's not really used, but I guess initializing it isn't such a bad idea anyway (not sure, but maybe `&iCharsWritten` could be changed to `nullptr` or simply 0 in that case, I don't think that crossed my mind back then...)

**Review by neico — CHANGES_REQUESTED** [review]
Left some nitpicks (I'd like to keep the code style consistent) as I can't validate if the code works fine with those changes and trust you on that and therefore focused on the actual diff's instead.

Also, while we're at it, maybe update the link reference at `DumpConsoleHelp()` to `https://github.com/EpicGames/UnrealEngine/blob/release/Engine/Source/Runtime/Engine/Private/UnrealEngine.cpp#L4260`?

**Comment by Scienziatogm** [comment]
Ok, I changed the code to follow your suggestions. I have tested this on Windows platform and everything works fine, while on Linux platform I'm getting some errors on which I'll work tomorrow. I haven't a Mac computer, so I cannot test the code on Mac platforms. Anyway after everything works I'll try to implement the Rcon feature and the autocompletion via tab feature.

**Comment by neico** [comment]
You missed a `FOutputDeviceConsolePlatform` at [ServerConsole.cpp](https://github.com/yas-online/UE4-DedicatedServer/pull/4/files#diff-75bef9b81fb98d7876bf9aae65810effR58)

**Comment by Scienziatogm** [comment]
> You missed a `FOutputDeviceConsolePlatform` at [ServerConsole.cpp](https://github.com/yas-online/UE4-DedicatedServer/pull/4/files#diff-75bef9b81fb98d7876bf9aae65810effR58)

Done in 82ec0eee2de09448b585a44590a598f590e4ac21

**Comment by Scienziatogm** [comment]
Seems like SetColor is a member of FOutputDeviceConsoleWindows, so it gives compile errros.

**Comment by Scienziatogm** [comment]
Everything seems to work now since I restored platform specific output devices, but today I've been working on the Linux console. When I'll finish, I'll make a commit. 😄

**Review by neico — APPROVED** [review]

**Comment by neico** [comment]
I think we can make a new PR for adding support for linux I guess

But this one seems ok so far, so I guess we can merge it?

**Comment by Scienziatogm** [comment]
> I think we can make a new PR for adding support for linux I guess
> 
> But this one seems ok so far, so I guess we can merge it?

Yes, if you want. When I finish the Linux support, I'll make another pull request.

**Pull request merged**